lib/libvfio-user.c

```diff
@@ -476,90 +476,102 @@ consume_fd(int *fds, size_t nr_fds, size_t index)
+
+    vfu_log(vfu_ctx, LOG_DEBUG, "adding DMA region %s", rstr);
+
+    if (region->flags == VFIO_USER_F_DMA_REGION_MAPPABLE) {
+        fd = consume_fd(msg->in_fds, msg->nr_in_fds, 0);
+        if (fd < 0) {
+            vfu_log(vfu_ctx, LOG_ERR,
+                    "failed to add DMA region %s: mappable but fd not provided",
+                    rstr);
+            return -1;
+        }
     }
 
-    nr_dma_regions = msg->in_size / sizeof(struct vfio_user_dma_region);
-    if (nr_dma_regions == 0) {
-        return 0;
+    ret = dma_controller_add_region(vfu_ctx->dma, (void *)region->addr,
+                                    region->size, fd, region->offset,
+                                    region->prot);
+    if (ret < 0) {
+        ret = errno;
+        vfu_log(vfu_ctx, LOG_ERR, "failed to add DMA region %s: %m", rstr);
+        if (fd != -1) {
+            close(fd);
+        }
+        return ERROR_INT(ret);
     }
 
-    for (i = 0, fdi = 0; i < nr_dma_regions; i++) {
-        struct vfio_user_dma_region *region = &dma_regions[i];
-        char rstr[1024];
+    if (vfu_ctx->dma_register != NULL) {
+        vfu_ctx->dma_register(vfu_ctx,
+                              &vfu_ctx->dma->regions[ret].info);
+    }
+    return 0;
+}
 
-        snprintf(rstr, sizeof(rstr), "[%#lx, %#lx) offset=%#lx "
-                "prot=%#x flags=%#x", region->addr, region->addr + region->size,
-                region->offset, region->prot, region->flags);
+int
+handle_dma_unmap(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg,
+                 struct vfio_user_dma_region *region)
+{
+    int ret;
 
-        vfu_log(vfu_ctx, LOG_DEBUG, "%s DMA region %s",
-                is_map ? "adding" : "removing", rstr);
+    assert(vfu_ctx != NULL);
+    assert(msg != NULL);
+    assert(region != NULL);
 
-        if (is_map) {
-            int fd = -1;
-            if (region->flags == VFIO_USER_F_DMA_REGION_MAPPABLE) {
-                fd = consume_fd(msg->in_fds, msg->nr_in_fds, fdi++);
-                if (fd < 0) {
-                    vfu_log(vfu_ctx, LOG_ERR, "failed to add DMA region %s: "
-                            "mappable but fd not provided", rstr);
-                    return -1;
-                }
-            }
+    if (msg->in_size < sizeof(*region)) {
+        vfu_log(vfu_ctx, LOG_ERR, "bad size of DMA unmap region %zu",
+                msg->in_size);
+        return ERROR_INT(EINVAL);
+    }
 
-            ret = dma_controller_add_region(vfu_ctx->dma, (void *)region->addr,
-                                            region->size, fd, region->offset,
-                                            region->prot);
-            if (ret < 0) {
-                ret = errno;
-                vfu_log(vfu_ctx, LOG_ERR, "failed to add DMA region %s: %m",
-                        rstr);
-                if (fd != -1) {
-                    close(fd);
-                }
-                return ERROR_INT(ret);
-            }
+    char rstr[1024];
```

Comment:
nit, I don't think we've been using C99-style here, should be at the top of the func

Reply:
right